### ISSUE TITLE
cant arrange brush sets

### ISSUE THREAD

**Issue by DarkDragon83** [mypaint/mypaint#550]
Hi,
on my laptop (Windows 10) mypaint works finde (i dont know the exact version, but i think its 1.1). Now i have a new wacon and want to install it on my desktop (windows 7, 64bit)
Everthing works fine, i first tried the 1.2 beta 4, install, started and wohooo works... until i tried to set up my workspace. Colours, layers, all works, but as i tried to move the brush sets to a position i like (left low corner), nothings works anymore. If i try to pull the brush window anywhere, it just lock up in the left top corner until i loose the mouse button, then i get a single window, which i cant move, cant close it, cant do anything with it, except creating more windows if i try to move it (then the same thin happen again and again). 
So i reinstalled - same problem, it even opens like in the screenshot. 
So deinstalled again, Restart my pc, and tried the 1.2 beta 3. But it opens quite the same, i cant do nothing... but i really would like to work with mypaint on my desktop :'(

if i use the "force a crash" option, it gives me this:

```python
Mypaint version: 1.2.0-beta.3+gitexport.0428646
System information: Windows-7-6.1.7601-SP1
Traceback (most recent call last):
  File "C:\Program Files\MyPaint\share\mypaint/gui/application.py", line 847, crash_program_cb(self=<gui.application.Application object>, action=<Action object at 0xfc5f7d0 (GtkAction at 0xf5d2850)>)
            """Tests exception handling."""
            raise Exception("This is a crash caused by the user.")
  variables: {'Exception': ('builtin', <type 'exceptions.Exception'>)}
Exception: This is a crash caused by the user.
```

Some help suggestions?

![mypaint](https://cloud.githubusercontent.com/assets/16556812/12113214/91bdd97c-b3a1-11e5-98a1-49eef7fbdd2c.jpg)

**Comment by achadwick** [comment]
Can you quickly drop a copy of your `settings.json` file into a gist (https://gist.github.com/) so I can look at it please? Feel free to edit it to remove any sensitive data!

I just want to look at the mess it's made of the windowing section. There should be no blank floating windows in there, particularly not ones which come back when MyPaint is closed and reopened!

It's a plain text file which can be edited in Notepad, then pasted. On windows, the `settings.json` file lives in `%LOCALAPPDATA%\mypaint`.

-----------------------

I can't reproduce this on Windows 7, but I have noticed that sometimes brush panel tabs cannot be dragged out of their bit of the sidebar. You'll be left with a GtkNotebook with just `[<] [blank] [>] [wrench] [x]` buttons and tabs in the top row and no content, and the sidebar ion the other side may still be open after you release the drag button.

The problem can be quickly resolved by clicking in the `[x]` button in the sidebar area you tried to drag the tab from.

(I think this is unrelated to yours though)

-----------------------

This is all very strange: there should be nothing weird about brush group tabs. Do you have any blank groups or other weirdness in your setup?

**Comment by achadwick** [comment]
What I suggest doing to make it workable again is to rename `settings.json` to something else (*after extracting its content to a gist, please!*). If you're feeling brave, you can edit it instead, and remove the `workspace.layout` settings. But please make a backup of your settings file first no matter what you do!

The setting block I'm talking about looks like

```json
  "workspace.layout": {
    "fullscreen": false, 
    "left_sidebar": {
      "h": 785, 
      "w": 267, 
      "groups": [
        {
          "tools": [], 
          "active_page": -1
        }
      ]
    }, 
    "autohide": true, 
    "maximized": false, 
    "position": {
      "y": 37, 
      "h": 887, 
      "w": 1048, 
      "x": 50
    }, 
    "floating": [], 
    "right_sidebar": {
      "h": 785, 
      "w": 21, 
      "groups": [
        {
          "tools": [], 
          "active_page": -1
        }
      ]
    }
  }, 
```

but it will be longer on your setup. Be very careful.

You'll need to remove the entire block, including the trailing comma if the curly brackets have one. Everything between the brackets in

```json
  "workspace.layout": {
    "fullscreen": false, 
    "left_sidebar": {
      "h": 785, 
      "w": 267,     "fullscreen": false, 
    "left_sidebar": {
      "h": 785, 
      "w": 267, 
      "groups": [
        {
          "tools": [], 
          "active_page": -1
        }
      ]
    }, 
    "autohide": true, 
    "maximized": false, 
    "position": {
      "y": 37, 
      "h": 887, 
      "w": 1048, 
      "x": 50
    }, 
    "floating": [], 
    "right_sidebar": {
      "h": 785, 
      "w": 21, 
      "groups": [
        {
          "tools": [], 
          "active_page": -1
        }
      ]
    }
  }, 
```

but it will be longer on your setup. It's simplest to convert it so something like

```json
  "workspace.layout": {
  }, 
```

then save `settings.json` and re-open MyPaint.

**Comment by DarkDragon83** [comment]
Ok, first i will copy you this file and dont change anything by myself
<script src="https://gist.github.com/DarkDragon83/3202611202b4e6706626.js"></script>

**Comment by achadwick** [comment]
Thanks for the saved config! I was able to reproduced the bad saved window layout under Linux, and yeah. Empty floating windows should be removed at startup to avoid them sitting there and generally being uncloseable and broken. That bit's fixed in 6eb969ed26b47c83cf8763df2c52a9cb97644e58.

So 1.2.0 final will silently weed out empty floating windows at startup now. Thanks for the info! If you want to test this fix now, you can find `gui/workspace.py` in your install and replace it with [the fixed gui/workspace.py](https://raw.githubusercontent.com/mypaint/mypaint/6eb969ed26b47c83cf8763df2c52a9cb97644e58/gui/workspace.py) - but it's really OK if you don't fancy mucking around with program code like that.

---------------------------------

I'm leaving this issue open so we can try to figure out how MyPaint got into the broken state in the first place. Have you been able to tidy up your `settings.json`, and does the issue come back when you do?

**Comment by DarkDragon83** [comment]
havent tried anything new yet, i will try at evening today

**Comment by DarkDragon83** [comment]
Hey there, 
so i tried to delete the workspace area, and IT WORKED!!! Wohooo...
But... after i tried to ajust my brush windows (at the last windows), i got the same bug again. So i deleted the section again, tried again and settled my brushes first (dont know if this was important), i got it and the other windows works without problem, just the brushes are crappy. 

Sooo i finished to settle my windows on my two screens just like i wanted (attached picture), and i really really dont want to reajust those windows in the future, so i have not to do this again ;)

![mypaint](https://cloud.githubusercontent.com/assets/16556812/12194435/48aef5f8-b5f0-11e5-9f35-8b67395c7dba.jpg)

But one little thing, i found a solution atm (but i think its not really a bug). If i close and reopen mypaint now, all my windows are still there, but crazy over my main screen, so i have to settle the windows on my second screen every time i open mypaint. Is there an option to fix them there?

thx a lot for your help

**Comment by achadwick** [comment]
Okay, looks like the main breakage in this issue has gone away for now, or has become unreproducable. I'll defer this issue for looking at after the 1.2.0 release is out.

We have a (fairly horrid) bit of code that's designed to detect and correct windows that have completely gone off-screen. Users removing monitors from their system is the usual cause of that. On Windows at least, we have to check each window MyPaint opens at startup, and throw it back onto the default monitor if it's offscreen. I bet this latest issue is related to that bit of code.

Grumpy developer opinion: floating windows like this are a pain in the proverbial when it comes to cross-platform support. I'm *so* tempted to make users pack things into sidebars or otherwise into the main window like Artrage's little tool buttons, but that hardly meets your use case. Looking at your screenshot, that makes sense too :neutral_face:

**Comment by DarkDragon83** [comment]
For me it makes a lot of sense, because i want my drawing space as huge as possible, but i dont want a cinema wall fitting screen (first, they are expansive and second they are a pain in the ass if you want to play on them too and cant sit 5 meters away). 
But i understand thats hard to program

**Comment by achadwick** [comment]
It looks like this may be fixed with the code commit in #596, which is superficially a very similar issue to the remnant of this one.

@DarkDragon83 
Could you take a look at #596's description please, and tell me if you think I can close the tail-end of this issue in favour of that one? Thanks.

**Comment by achadwick** [comment]
The fix above is now tested and backported ready for 1.2.1.

**Comment by achadwick** [comment]
No feedback, so I'm tentatively marking this one as closed now that 95f843e750cb0e0547e6e535edc9b3fca6b665cd (the "fix above" mentioned earlier) has been pushed for v1.2.1.

**Comment by Botimes** [comment]
I have a workaround!  You can select each window and close them with Alt+Tab!!
